add a page for OfAS permissions (#4161)
spotinst/help: src/docs/ocean-spark/configure-permissions/README.md

# Configure Permissions

This section details how to setup permissions for users of the **Ocean for Apache Spark** product.

> If you're not familiar with how permissions work in Spot, we recommend you take a look at this section : [Permission Policies](/administration/policies)

Reviewer: ```suggestion
If you're not familiar with how permissions work in Spot, we recommend you take a look at this section: [Permission Policies](/administration/policies)
```

## Overview

> Ocean for Apache Spark let you configure permission policies for **Actions** for the following resource types:

Reviewer: ```suggestion
Ocean for Apache Spark lets you configure permission policies for **Actions** for the following resource types:
```

- Clusters
- Apps
- Jobs

> - Config-templates

Reviewer: ```suggestion
- Configuration templates
```

- Workspaces

All these resource types at least bound to a **Cluster** resource.
In Ocean for Apache Spark, clusters are identified by an id with the format `osc-xxxxxxxx`

## Account-level Managed Permission Policies

To get started, you can use one of the following managed policies.  

> They will apply at the account level. If you want to set permissions more granular, look at [Granular Permissions Policies How-to Guides section](#granular-permissions-policies-how-to-guides)

Reviewer: ```suggestion
They will apply at the account level. If you want to set permissions with more granularity, look at the [Granular Permissions Policies How-to Guides section](#granular-permissions-policies-how-to-guides)
```

| Policy                                | Level        | Effect                                                    | Product Scope                                      | Ocean Spark Resource Scope                                                                      |
| ------------------------------------- | ------------ | --------------------------------------------------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------------------- |
| Account Viewer                        | Spot account | Give read access                                          | All Spot products including Ocean for Apache Spark | All Resources                                                                                   |
| Account Editor                        | Spot account | Give edit access<br>**including app submission**          | All Spot products including Ocean for Apache Spark | All Resources                                                                                   |

## Granular Permissions Policies How-to Guides

> Patterns describe below are for "edit" actions only.

Reviewer: ```suggestion
Patterns described below are for "edit" actions only.
```

> Users should have a read-access managed policies attached to their profile.

Reviewer: ```suggestion
Users should have read-access managed policies attached to their profile.
```

| Account Viewer                        | Spot account | Give read access                                          | All Spot products including Ocean for Apache Spark | All Resources                                                                                   |
| Account Editor                        | Spot account | Give edit access<br>**including app submission**          | All Spot products including Ocean for Apache Spark | All Resources                                                                                   |

## Granular Permissions Policies How-to Guides

Patterns described below are for "edit" actions only.
Users should have read-access managed policies attached to their profile.

### How to define permissions for one or several clusters

Cluster is the default resource to scope a permission policy.
You can do it by using the `resources` field in the policy :

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": ["spark:*"],
      "resources": [
        "sparkClusterId:osc-cluster1",
        "sparkClusterId:osc-cluster2",
        "sparkConfigTemplateId:*",
        "sparkJobId:*"
      ]
    }
  ]
}
```

> ### How to define permissions for a specific set of config-template

Reviewer: ```suggestion
### How to define permissions for a specific set of config-templates
```

You can specify a clusterId and several configTemplateId like this :

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": ["spark:*"],
      "resources": [
        "sparkClusterId:osc-cluster1",
        "sparkConfigTemplateId:config-template-1",
        "sparkConfigTemplateId:config-template-2",
        "sparkJobId:*"
      ]
    }
  ]
}
```

You can also use a wildcard `*` in the config-template resource value like :

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": ["spark:*"],
      "resources": [
        "sparkClusterId:osc-cluster1",
        "sparkConfigTemplateId:config-template-*-abc-*",
        "sparkJobId:*"
      ]
    }
  ]
}
```

> **Important Note 1:** If you specify several clusters and several config-templates,
> it will allow users to do operations related to any config-template specified against any cluster specified, even if you declare several `statements` with different `resources` array

> > **Important Note 2:** If you want to force users to use a `configTemplateId` at app submisssion, you need to use the `condition` field as described below

Reviewer: ```suggestion
> **Important Note 2:** If you want to force users to use a `configTemplateId` in app submissions, you need to use the `condition` field as described below
```

> If you wish to only allow operations for a specific set of config-templates against a specific clusters, you'll need to use `condition` field.

Reviewer: ```suggestion
If you wish to only allow operations for a specific set of config-templates against a specific set of clusters, you'll need to use the `condition` field.
```

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": ["spark:*"],
      "resources": ["*"],
      "condition": {
        "Or": [
          {
            "And": [
              {
                "Or": [
                  {
                    "StringEquals": {
                      "sparkClusterId": "osc-cluster1"
                    }
                  },
                  {
                    "StringEquals": {
                      "sparkClusterId": "osc-cluster2"
                    }
                  }
                ]
              },
              {
                "StringPatternMatch": {
                  "sparkConfigTemplateId": "ct-team-a-*"
                }
              }
            ]
          },
          {
            "And": [
              {
                "StringEquals": {
                  "sparkClusterId": "osc-cluster1"
                }
              },
              {
                "StringPatternMatch": {
                  "sparkConfigTemplateId": "ct-notebook-*"
                }
              }
            ]
          }
        ]
      }
    }
  ]
}
```

## Use cases

### Set permissions to isolate app submission by team

> Let's say you want each of your team to have their own config-templates and to be able to only submit spark application for their config-template.

Reviewer: ```suggestion
Let's say you want each of your teams to have their own config-templates and to be able to only submit Spark applications using their config-templates.
```

For each team, you can:

> 1. Create config-template using a specific pattern for the id

Reviewer: ```suggestion
1. Create config-templates using a specific pattern for the id
```

   For example : `team-A-config-template-1`, `team-A-config-template-2`, `team-B-config-template-1`

2. Attach a managed policy like `Account Viewer` to your users.

> 3. Create and attach one of the following policy by team:

Reviewer: ```suggestion
3. Create and attach one of the following policies by team:
```

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication", // optional
        "spark:createConfigTemplate",
        "spark:updateConfigTemplate",
        "spark:deleteConfigTemplate"
      ],
      "resources": ["sparkClusterId:*", "sparkConfigTemplateId:team-a*"]
    }
  ]
}
```

> If you want to also restrict the cluster users are allowed to submit to:

Reviewer: ```suggestion
If you want to also restrict the cluster users are allowed to submit to:
If you also want to restrict which cluster users are allowed to submit applications to:
```

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication", // optional
        "spark:createConfigTemplate",
        "spark:updateConfigTemplate",
        "spark:deleteConfigTemplate"
      ],
      "resources": [
        "sparkClusterId:osc-xxxxxx",
        "sparkConfigTemplateId:team-a*"
      ]
    }
  ]
}
```

If you want to **force users to use a config-template when submitting an app**, you need to use the `condition` field like so:

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication", // optional
        "spark:createConfigTemplate",
        "spark:updateConfigTemplate",
        "spark:deleteConfigTemplate"
      ],
      "resources": ["*"],
      "condition": {
        "And": [
          {
            "StringEquals": {
              "sparkClusterId": "osc-xxxxxx"
            },
            "StringPatternMatch": {
              "sparkConfigTemplateId": "ct-team-a-*"
            }
          }
        ]
      }
    }
  ]
}
```

### Set permissions for notebook users

> If you want to give your users only access to notebook feature using local Jupyter notebook or local JupyterLab instance,

Reviewer: ```suggestion
If you want to give your users access to the notebook feature using local Jupyter notebooks or a JupyterLab instance,
```

you can use the following policy

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication",
        "spark:createNotebook",
        "spark:updateNotebook",
        "spark:deleteNotebook"
      ],
      "resources": ["*"]
    }
  ]
}
```

> If you want to allow notebook use only for a subset of config-template, you can have something like :

Reviewer: ```suggestion
If you want to allow notebook use only for a subset of config-templates, you can have something like:
```

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication",
        "spark:createNotebook",
        "spark:updateNotebook",
        "spark:deleteNotebook"
      ],
      "resources": [
        "sparkClusterId:osc-xxxxxx",
        "sparkConfigTemplateId:team-a*"
      ]
    }
  ]
}
```

### Set permissions for workspace users

> If you want to give your users only access to integrated workspace feature,

Reviewer: ```suggestion
If you want to give your users only access to the integrated notebook workspace feature,
```

you can use the following policy

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication",
        "spark:createNotebook",
        "spark:updateNotebook",
        "spark:deleteNotebook",
        "spark:createWorkspace",
        "spark:updateWorkspace",
        "spark:deleteWorkspace",
        "spark:createWorkspaceProxy",
        "spark:updateWorkspaceProxy",
        "spark:deleteWorkspaceProxy"
      ],
      "resources": ["*"]
    }
  ]
}
```

> If you want to allow workspace use only for a subset of config-template, you can have something like :

Reviewer: ```suggestion
If you want to allow workspace use only for a subset of config-templates, you can have something like:
```

```json
{
  "statements": [
    {
      "effect": "ALLOW",
      "actions": [
        "spark:createApplication",
        "spark:deleteApplication",
        "spark:createNotebook",
        "spark:updateNotebook",
        "spark:deleteNotebook",
        "spark:createWorkspace",
        "spark:updateWorkspace",
        "spark:deleteWorkspace",
        "spark:createWorkspaceProxy",
        "spark:updateWorkspaceProxy",
        "spark:deleteWorkspaceProxy"
      ],
      "resources": [
        "sparkClusterId:osc-xxxxxx",
        "sparkConfigTemplateId:team-a*"
      ]
    }
  ]
}
```

## Advanced Policy Patterns

> If you want more complex rules and combination between resources, you can use the `condition` field,

Reviewer: ```suggestion
If you want more complex rules and combinations between resources, you can use the `condition` field,
```

you can learn more about it here: [Policy conditions](/administration/policies/create-new-policy?id=policy-conditions)

## Reference

> All actions are listed below are "edit" permissions

Reviewer: ```suggestion
All actions listed below are "edit" permissions
```
